src/modules/mavlink/mavlink_main.cpp

```diff
@@ -1217,117 +1217,12 @@ Mavlink::configure_stream_threadsafe(const char *stream_name, const float rate)
 	}
 }
 
-int
-Mavlink::message_buffer_init(int size)
-{
-	_message_buffer.size = size;
-	_message_buffer.write_ptr = 0;
-	_message_buffer.read_ptr = 0;
-	_message_buffer.data = (char *)malloc(_message_buffer.size);
-
-	int ret;
-
-	if (_message_buffer.data == nullptr) {
-		ret = PX4_ERROR;
-		_message_buffer.size = 0;
-
-	} else {
-		ret = OK;
-	}
-
-	return ret;
-}
-
-void
-Mavlink::message_buffer_destroy()
-{
-	_message_buffer.size = 0;
-	_message_buffer.write_ptr = 0;
-	_message_buffer.read_ptr = 0;
-	free(_message_buffer.data);
-}
-
-int
-Mavlink::message_buffer_count()
-{
-	int n = _message_buffer.write_ptr - _message_buffer.read_ptr;
-
-	if (n < 0) {
-		n += _message_buffer.size;
-	}
-
-	return n;
-}
-
-bool
-Mavlink::message_buffer_write(const void *ptr, int size)
-{
-	// bytes available to write
-	int available = _message_buffer.read_ptr - _message_buffer.write_ptr - 1;
-
-	if (available < 0) {
-		available += _message_buffer.size;
-	}
-
-	if (size > available) {
-		// buffer overflow
-		return false;
-	}
-
-	char *c = (char *) ptr;
-	int n = _message_buffer.size - _message_buffer.write_ptr;	// bytes to end of the buffer
-
-	if (n < size) {
-		// message goes over end of the buffer
-		memcpy(&(_message_buffer.data[_message_buffer.write_ptr]), c, n);
-		_message_buffer.write_ptr = 0;
-
-	} else {
-		n = 0;
-	}
-
-	// now: n = bytes already written
-	int p = size - n;	// number of bytes to write
-	memcpy(&(_message_buffer.data[_message_buffer.write_ptr]), &(c[n]), p);
-	_message_buffer.write_ptr = (_message_buffer.write_ptr + p) % _message_buffer.size;
-	return true;
-}
-
-int
-Mavlink::message_buffer_get_ptr(void **ptr, bool *is_part)
-{
-	// bytes available to read
-	int available = _message_buffer.write_ptr - _message_buffer.read_ptr;
-
-	if (available == 0) {
-		return 0;	// buffer is empty
-	}
-
-	int n = 0;
-
-	if (available > 0) {
-		// read pointer is before write pointer, all available bytes can be read
-		n = available;
-		*is_part = false;
-
-	} else {
-		// read pointer is after write pointer, read bytes from read_ptr to end of the buffer
-		n = _message_buffer.size - _message_buffer.read_ptr;
-		*is_part = _message_buffer.write_ptr > 0;
-	}
-
-	*ptr = &(_message_buffer.data[_message_buffer.read_ptr]);
-	return n;
-}
-
 void
 Mavlink::pass_message(const mavlink_message_t *msg)
 {
-	/* size is 8 bytes plus variable payload */
+	/* size is 12 bytes plus variable payload */
 	int size = MAVLINK_NUM_NON_PAYLOAD_BYTES + msg->len;
-	pthread_mutex_lock(&_message_buffer_mutex);
-	message_buffer_write(msg, size);
-	pthread_mutex_unlock(&_message_buffer_mutex);
+	_message_buffer.push_back(reinterpret_cast<const uint8_t *>(msg), size);
```

Comment:
It would be nice to track failures (no space available) so that something could actually be done about it later.

Reply:
Ok, I'll add a perf counter.